File: android-twa/app/src/main/java/com/akolta/recruitpro/MainActivity.java
package com.akolta.recruitpro;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Window;
import android.view.WindowManager;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.webkit.GeolocationPermissions;

/**
 * RecruitPro Android WebView — Native calling bridge (v1.2.0).
 *
 * <p>JavaScript bridge ("AndroidBridge") exposed to the web app:
 * <ul>
 *   <li>makeCall(phoneNumber)          — ACTION_CALL (direct dial, no dialer UI)</li>
 *   <li>hasCallPermission()            — returns "true"/"false" for runtime check</li>
 *   <li>requestCallPermission()        — proactively request CALL_PHONE permission</li>
 *   <li>openAppSettings()              — open app settings for permission management</li>
 * </ul>
 *
 * <p>Callbacks to web app:
 * <ul>
 *   <li>onCallResult(resultCode, message) — notifies web of call outcome</li>
 *   <li>showPostCallDisposition('')        — fires when user returns from a placed call</li>
 * </ul>
 *
 * <p>Call result codes:
 * <ul>
 *   <li>CALL_INITIATED               — Call was successfully initiated via ACTION_CALL</li>
 *   <li>PERMISSION_DENIED            — User denied permission (can ask again)</li>
 *   <li>PERMISSION_PERMANENTLY_DENIED — User permanently denied (must open Settings)</li>
 *   <li>CALL_FAILED                  — Call initiation failed (no SIM, carrier issue, etc.)</li>
 *   <li>INVALID_NUMBER               — Phone number is invalid</li>
 * </ul>
 */
public class MainActivity extends Activity {

    private static final int REQUEST_CALL_PHONE = 1001;
    private WebView webView;
    private String pendingCallNumber = null;
    private boolean callPlaced = false;

    // ─────────────────────────────────────────────────────────────────
    // JavaScript Interface — exposed as window.AndroidBridge
    // ─────────────────────────────────────────────────────────────────
    private class AndroidBridge {

        /**
         * Place a direct phone call (ACTION_CALL).
         * Requests runtime permission if not yet granted.
         * Notifies the web app of the result via onCallResult callback.
         *
         * @param phoneNumber The phone number to call (digits, optional leading +)
         */
        @JavascriptInterface
        public void makeCall(String phoneNumber) {
            if (phoneNumber == null || phoneNumber.isEmpty()) {
                notifyCallResult("INVALID_NUMBER", "Phone number is empty");
                return;
            }

            // Clean the number — keep digits and leading +
            String clean = phoneNumber.replaceAll("[^0-9+]", "");
            if (clean.length() < 7) {
                notifyCallResult("INVALID_NUMBER", "Phone number is too short (minimum 7 digits)");
                return;
            }

            runOnUiThread(() -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
                        && checkSelfPermission(Manifest.permission.CALL_PHONE)
                        != PackageManager.PERMISSION_GRANTED) {
                    // Permission not yet granted — request it
                    pendingCallNumber = clean;
                    requestPermissions(
                            new String[]{Manifest.permission.CALL_PHONE},
                            REQUEST_CALL_PHONE);
                    return;
                }

                // Permission granted — direct call via ACTION_CALL
                placeCall(clean);
            });
        }

        /**
         * Returns "true" if CALL_PHONE runtime permission is granted.
         */
        @JavascriptInterface
        public String hasCallPermission() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                return checkSelfPermission(Manifest.permission.CALL_PHONE)
                        == PackageManager.PERMISSION_GRANTED ? "true" : "false";
            }
            // Pre-Marshmallow: permission is granted at install time
            return "true";
        }

        /**
         * Proactively request CALL_PHONE permission.
         * The result is delivered via onCallResult callback.
         */
        @JavascriptInterface
        public void requestCallPermission() {
            runOnUiThread(() -> {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    if (checkSelfPermission(Manifest.permission.CALL_PHONE)
                            == PackageManager.PERMISSION_GRANTED) {
                        // Already granted
                        notifyCallResult("CALL_INITIATED", "Permission already granted");
                    } else {
                        requestPermissions(
                                new String[]{Manifest.permission.CALL_PHONE},
                                REQUEST_CALL_PHONE);
                    }
                } else {
                    // Pre-Marshmallow: permission is granted at install time
                    notifyCallResult("CALL_INITIATED", "Permission already granted");
                }
            });
        }

        /**
         * Open this app's system Settings page.
         * Used when CALL_PHONE permission is permanently denied.
         */
        @JavascriptInterface
        public void openAppSettings() {
            runOnUiThread(() -> {
                try {
                    Intent intent = new Intent(
                            Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                            Uri.parse("package:" + getPackageName()));
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            });
        }
    }

    // ─────────────────────────────────────────────────────────────────
    // Lifecycle
    // ─────────────────────────────────────────────────────────────────
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setStatusBarColor(Color.parseColor("#047857"));
        getWindow().setNavigationBarColor(Color.parseColor("#059669"));

        webView = new WebView(this);
        setContentView(webView);

        // ── WebView settings ─────────────────────────────────────────
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setDatabaseEnabled(true);
        settings.setAllowFileAccess(true);
        settings.setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        settings.setCacheMode(WebSettings.LOAD_DEFAULT);
        settings.setUseWideViewPort(true);
        settings.setLoadWithOverviewMode(true);
        settings.setSupportZoom(false);
        settings.setMediaPlaybackRequiresUserGesture(false);
        settings.setGeolocationEnabled(true);

        // ── Expose JavaScript bridge ─────────────────────────────────
        webView.addJavascriptInterface(new AndroidBridge(), "AndroidBridge");

        // ── URL interception ─────────────────────────────────────────
        webView.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                if (url == null) return false;

                // Allow app's own URLs to load inside WebView
                if (url.startsWith("https://app.akolta.com")) {
                    return false;
                }

                // tel: — let the AndroidBridge.makeCall handle direct calling.
                // If somehow a raw tel: link is clicked (not via bridge),
                // fall back to ACTION_DIAL (opens dialer).
                if (url.startsWith("tel:")) {
                    try {
                        String number = url.substring(4).trim();
                        Intent intent = new Intent(Intent.ACTION_DIAL,
                                Uri.parse("tel:" + number));
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    return true;
                }

                // sms: — open native SMS app
                if (url.startsWith("sms:")) {
                    try {
                        Intent intent = new Intent(Intent.ACTION_VIEW,
                                Uri.parse(url));
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    return true;
                }

                // whatsapp: — open WhatsApp
                if (url.startsWith("whatsapp://")) {
                    try {
                        Intent intent = new Intent(Intent.ACTION_VIEW,
                                Uri.parse(url));
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                        startActivity(intent);
                    } catch (Exception e) {
                        // WhatsApp not installed — try Play Store
                        try {
                            Intent store = new Intent(Intent.ACTION_VIEW,
                                    Uri.parse("https://play.google.com/store/apps/details?id=com.whatsapp"));
                            startActivity(store);
                        } catch (Exception ignored) {}
                    }
                    return true;
                }

                // All other external URLs — open in system browser
                try {
                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return true;
            }
        });

        // ── Chrome client (geolocation, etc.) ────────────────────────
        webView.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onGeolocationPermissionsShowPrompt(String origin,
                    GeolocationPermissions.Callback callback) {
                callback.invoke(origin, true, false);
            }
        });

        webView.loadUrl("https://app.akolta.com/");
    }

    // ─────────────────────────────────────────────────────────────────
    // Return-from-call detection
    // ─────────────────────────────────────────────────────────────────
    @Override
    protected void onResume() {
        super.onResume();
        if (webView != null) {
            webView.onResume();

            // Only notify the web app if a call was actually placed.
            // This prevents false disposition triggers when:
            //   - Permission was denied (no call was made)
            //   - User temporarily left the app for other reasons
            if (callPlaced) {
                callPlaced = false;
                webView.postDelayed(() -> {
                    try {
                        webView.evaluateJavascript(
                                "if(typeof showPostCallDisposition==='function') showPostCallDisposition('');",
                                null
                        );
                    } catch (Exception e) {
                        // WebView not ready yet — ignore
                    }
                }, 300);
            }
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (webView != null) {
            webView.onPause();
        }
    }

    @Override
    public void onBackPressed() {
        if (webView != null && webView.canGoBack()) {
            webView.goBack();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onDestroy() {
        if (webView != null) {
            webView.destroy();
        }
        super.onDestroy();
    }

    // ─────────────────────────────────────────────────────────────────
    // Runtime permission result
    // ─────────────────────────────────────────────────────────────────
    @Override
    public void onRequestPermissionsResult(int requestCode,
            String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == REQUEST_CALL_PHONE) {
            if (grantResults.length > 0
                    && grantResults[0] == PackageManager.PERMISSION_GRANTED
                    && pendingCallNumber != null) {
                // Permission granted — place the pending call
                String number = pendingCallNumber;
                pendingCallNumber = null;
                placeCall(number);
            } else {
                // Permission denied — determine if permanently denied
                pendingCallNumber = null;
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    if (shouldShowRequestPermissionRationale(Manifest.permission.CALL_PHONE)) {
                        // User denied but can be asked again
                        notifyCallResult("PERMISSION_DENIED",
                                "Phone call permission is required to place calls directly from Akolta Dialer.");
                    } else {
                        // User checked "Don't ask again" or this is the first denial
                        // on a device that doesn't show rationale before first request.
                        // Treat as permanently denied — user must go to Settings.
                        notifyCallResult("PERMISSION_PERMANENTLY_DENIED",
                                "Phone call permission has been permanently denied. "
                                + "Please enable it in App Settings to place calls directly.");
                    }
                } else {
                    notifyCallResult("PERMISSION_DENIED",
                            "Phone call permission is required.");
                }
            }
        }
    }

    // ─────────────────────────────────────────────────────────────────
    // Helper: place a call using ACTION_CALL (direct, no dialer UI)
    // ─────────────────────────────────────────────────────────────────
    private void placeCall(String phoneNumber) {
        try {
            // ACTION_CALL = direct call (no dialer UI)
            // Requires CALL_PHONE permission (declared in AndroidManifest + runtime grant)
            Intent callIntent = new Intent(Intent.ACTION_CALL,
                    Uri.parse("tel:" + phoneNumber));
            callIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            startActivity(callIntent);

            // Mark that a call was placed — onResume will use this to
            // fire showPostCallDisposition when the user returns.
            callPlaced = true;

            // Notify web app that the call was initiated successfully
            notifyCallResult("CALL_INITIATED", "Call placed to " + phoneNumber);

        } catch (SecurityException e) {
            // ACTION_CALL failed due to missing permission
            callPlaced = false;
            notifyCallResult("PERMISSION_PERMANENTLY_DENIED",
                    "Permission denied. Please grant phone call permission in App Settings.");
        } catch (android.content.ActivityNotFoundException e) {
            // No telephony/calling app available on the device
            callPlaced = false;
            notifyCallResult("CALL_FAILED",
                    "No calling app available on this device. Please ensure a SIM card is inserted.");
        } catch (Exception e) {
            callPlaced = false;
            String msg = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
            if (msg.contains("no sim") || msg.contains("no_sim")
                    || msg.contains("not attached") || msg.contains("call_phone")
                    || msg.contains("not available")) {
                notifyCallResult("CALL_FAILED",
                        "Unable to place call. Please check if a SIM card is inserted "
                        + "and your device can make calls.");
            } else {
                notifyCallResult("CALL_FAILED",
                        "Failed to place call: " + (e.getMessage() != null ? e.getMessage() : "Unknown error"));
            }
        }
    }

    // ─────────────────────────────────────────────────────────────────
    // Helper: send a callback to the web app
    // ─────────────────────────────────────────────────────────────────
    private void notifyCallResult(final String resultCode, final String message) {
        if (webView == null) return;

        // Escape special characters for safe JavaScript string embedding
        String escaped = message != null
                ? message.replace("\\", "\\\\")
                        .replace("\"", "\\\"")
                        .replace("\'", "\\'")
                        .replace("\n", "\\n")
                        .replace("\r", "\\r")
                : "";

        final String js = String.format(
                "if(typeof onCallResult==='function') onCallResult(\"%s\",\"%s\");",
                resultCode, escaped);

        runOnUiThread(() -> {
            try {
                webView.evaluateJavascript(js, null);
            } catch (Exception e) {
                // WebView not ready — ignore silently
            }
        });
    }
}

File: android-files/MainActivity.kt
package com.yourpackage.app

import android.Manifest
import android.annotation.SuppressLint
import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.speech.RecognizerIntent
import android.util.Log
import android.view.View
import android.view.WindowInsets
import android.view.WindowInsetsControllerCompat
import android.view.WindowManager
import android.webkit.ConsoleMessage
import android.webkit.DownloadListener
import android.webkit.PermissionRequest
import android.webkit.URLUtil
import android.webkit.WebChromeClient
import android.webkit.WebResourceRequest
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import java.io.File


/**
 * ===================================================================
 * COMPLETE MainActivity.kt — SINGLE FILE, NO OTHER FILES NEEDED
 * ===================================================================
 *
 * HOW TO USE THIS FILE:
 * 1. Open your Android Studio project
 * 2. In the LEFT panel, go to: app > java > com > yourcompany > yourapp
 * 3. Open MainActivity.kt (double-click it)
 * 4. SELECT ALL text in that file (Ctrl+A)
 * 5. DELETE it (Delete key)
 * 6. COPY this ENTIRE file content and PASTE it
 * 7. Change "com.yourpackage.app" on line 1 to YOUR package name
 *    (see SETUP-GUIDE.txt for how to find your package name)
 * 8. Change "https://your-app-url.com" on line ~70 to YOUR web app URL
 * 9. Click Build > Rebuild Project
 * 10. Click the green Run button
 *
 * WHAT THIS FILE DOES:
 * - Opens phone dialer when user taps "Call" button in the web app
 * - Detects when user returns from phone call
 * - Shows Post Call Disposition popup automatically
 * - Does NOT reload the web page when returning from dialer
 * - Enables full-screen immersive mode for maximum viewport
 * - Injects viewport fix JavaScript on every page load
 * - Provides native speech recognition via Android Bridge
 * - Handles SMS and WhatsApp intent links
 * - Requests RECORD_AUDIO permission at runtime (Android 6.0+)
 * ===================================================================
 */
class MainActivity : AppCompatActivity() {

    private lateinit var webView: WebView
    private val handler = Handler(Looper.getMainLooper())

    // Modern ActivityResultLauncher for speech recognition (replaces deprecated startActivityForResult)
    private lateinit var speechLauncher: ActivityResultLauncher<Intent>
    // Modern ActivityResultLauncher for runtime permissions (RECORD_AUDIO)
    private lateinit var permissionLauncher: ActivityResultLauncher<String>

    // Flag to remember we need to start speech after permission is granted
    private var pendingSpeechAfterPermission = false

    // =============================================================
    //   CHANGE THIS to your actual web app URL
    //   Example: "https://recruitpro.example.com"
    // =============================================================
    private val APP_URL = "https://your-app-url.com"

    // Track if phone dialer was opened (used in onResume)
    private var isCallInProgress = false
    private var currentCallPhone = ""

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // ========== STEP 0: Enable fullscreen immersive mode ==========
        enableImmersiveMode()

        // ========== STEP 0.5: Register ActivityResultLaunchers ==========
        // Speech recognition result launcher
        speechLauncher = registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result ->
            if (result.resultCode == RESULT_OK && result.data != null) {
                val results = result.data?.getStringArrayListExtra(RecognizerIntent.EXTRA_RESULTS)
                val transcript = results?.firstOrNull() ?: ""
                val escaped = transcript
                    .replace("\\", "\\\\")
                    .replace("'", "\\'")
                    .replace("\n", "\\n")
                Log.d("[Voice]", "Speech result: $transcript")
                webView.evaluateJavascript("window.__onSpeechResult('$escaped', null);", null)
            } else {
                Log.d("[Voice]", "Speech cancelled (resultCode=${result.resultCode})")
                webView.evaluateJavascript("window.__onSpeechResult(null, 'cancelled');", null)
            }
        }

        // Runtime permission launcher for RECORD_AUDIO
        permissionLauncher = registerForActivityResult(
            ActivityResultContracts.RequestPermission()
        ) { granted ->
            if (granted) {
                Log.d("[Voice]", "RECORD_AUDIO permission granted")
                launchSpeechRecognizer()
            } else {
                Log.d("[Voice]", "RECORD_AUDIO permission denied")
                webView.evaluateJavascript("window.__onSpeechResult(null, 'permission_denied');", null)
                pendingSpeechAfterPermission = false
            }
        }

        // ========== STEP 1: Create WebView ==========
        webView = WebView(this)
        setContentView(webView)

        // ========== STEP 2: Enable JavaScript & Settings ==========
        webView.settings.apply {
            javaScriptEnabled = true
            domStorageEnabled = true
            allowFileAccess = true
            allowContentAccess = true
            mediaPlaybackRequiresUserGesture = false
            cacheMode = WebSettings.LOAD_DEFAULT
            mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
            useWideViewPort = true
            loadWithOverviewMode = true
            setSupportMultipleWindows(false)

            // --- IMPORTANT: Viewport & scroll fixes for bottom-sheet popups ---
            setLayoutAlgorithm(WebSettings.LayoutAlgorithm.TEXT_AUTOSIZING)
            setBuiltInZoomControls(false)
            displayZoomControls = false
        }

        // --- Enable smooth scrolling ---
        webView.isVerticalScrollBarEnabled = true
        webView.isHorizontalScrollBarEnabled = false
        webView.overScrollMode = WebView.OVER_SCROLL_NEVER

        // ========== STEP 2.5: Handle file downloads (Excel, PDF, etc.) ==========
        webView.setDownloadListener(object : DownloadListener {
            override fun onDownloadStart(
                url: String?,
                userAgent: String?,
                contentDisposition: String?,
                mimetype: String?,
                contentLength: Long
            ) {
                if (url == null) {
                    Log.e("[Download]", "Download URL is null")
                    return
                }

                Log.d("[Download]", "Download requested: $url")

                try {
                    val fileName = contentDisposition?.let {
                        val regex = Regex("filename\\*?=(?:\"|UTF-8'')(.*?)(?:\"|$)")
                        regex.find(it)?.groupValues?.get(1)?.trim()
                    } ?: URLUtil.guessFileName(url, contentDisposition, mimetype)

                    Log.d("[Download]", "Saving as: $fileName")

                    val request = DownloadManager.Request(Uri.parse(url)).apply {
                        setMimeType(mimetype)
                        addRequestHeader("User-Agent", userAgent)
                        setDescription("Downloading $fileName")
                        setTitle(fileName)
                        setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName)
                        setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
                        allowScanningByMediaScanner()
                    }

                    val dm = getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
                    dm.enqueue(request)
                    Log.d("[Download]", "Download enqueued: $fileName")
                } catch (e: Exception) {
                    Log.e("[Download]", "Download FAILED: ${e.message}")
                    try {
                        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url)).apply {
                            flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        }
                        startActivity(intent)
                    } catch (e2: Exception) {
                        Log.e("[Download]", "Browser fallback also FAILED: ${e2.message}")
                    }
                }
            }
        })

        // ========== STEP 3: WebViewClient — handles tel:/sms:/wa.me links + viewport fix ==========
        webView.webViewClient = object : WebViewClient() {

            override fun shouldOverrideUrlLoading(
                view: WebView?,
                request: WebResourceRequest?
            ): Boolean {
                val url = request?.url?.toString() ?: return false

                // ---------- HANDLE tel: LINKS (Phone Dialer) ----------
                if (url.startsWith("tel:")) {
                    val phone = url.removePrefix("tel:")
                    isCallInProgress = true
                    currentCallPhone = phone

                    Log.d("[Dialer]", "Opening phone dialer for: $phone")

                    try {
                        val dialIntent = Intent(Intent.ACTION_DIAL).apply {
                            data = Uri.parse(url)
                            flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        }
                        startActivity(dialIntent)
                    } catch (e: Exception) {
                        Log.e("[Dialer]", "FAILED to open dialer: ${e.message}")
                        isCallInProgress = false
                        currentCallPhone = ""
                    }
                    return true
                }

                // ---------- HANDLE sms: LINKS (SMS App) ----------
                if (url.startsWith("sms:")) {
                    Log.d("[Dialer]", "Opening SMS for: $url")
                    try {
                        val smsIntent = Intent(Intent.ACTION_SENDTO).apply {
                            data = Uri.parse(url)
                            flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        }
                        startActivity(smsIntent)
                    } catch (e: Exception) {
                        Log.e("[Dialer]", "FAILED to open SMS: ${e.message}")
                    }
                    return true
                }

                // ---------- HANDLE WhatsApp links ----------
                if (url.contains("wa.me") || url.startsWith("whatsapp://")) {
                    Log.d("[Dialer]", "Opening WhatsApp: $url")
                    try {
                        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url)).apply {
                            flags = Intent.FLAG_ACTIVITY_NEW_TASK
                        }
                        startActivity(intent)
                    } catch (e: Exception) {
                        Log.e("[Dialer]", "FAILED to open WhatsApp: ${e.message}")
                    }
                    return true
                }

                return false
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                Log.d("[Dialer]", "Page loaded: $url")

                // Inject JavaScript to fix viewport height for bottom-sheet popups.
                view?.evaluateJavascript("""
                    (function() {
                        setTimeout(function() {
                            window.dispatchEvent(new Event('resize'));
                            console.log('[ViewportFix] Resize dispatched, innerHeight=' + window.innerHeight);
                        }, 300);
                        setTimeout(function() {
                            window.dispatchEvent(new Event('resize'));
                            console.log('[ViewportFix] Second resize, innerHeight=' + window.innerHeight);
                        }, 1000);
                    })();
                """.trimIndent(), null)

                // After page loads, if we have a pending call, trigger disposition
                if (isCallInProgress && currentCallPhone.isNotEmpty()) {
                    handler.postDelayed({
                        triggerDispositionPopup(currentCallPhone)
                    }, 800)
                }
            }
        }

        // ========== STEP 4: WebChromeClient (console.log + permission requests) ==========
        webView.webChromeClient = object : WebChromeClient() {
            override fun onConsoleMessage(message: ConsoleMessage?): Boolean {
                Log.d("[WebView]", "${message?.message()}")
                return true
            }

            // Handle WebView-internal permission requests (microphone for getUserMedia)
            override fun onPermissionRequest(request: PermissionRequest?) {
                if (request != null) {
                    if (request.resources.contains(PermissionRequest.RESOURCE_AUDIO_CAPTURE)) {
                        Log.d("[WebView]", "Granting microphone permission for getUserMedia")
                        request.grant(request.resources)
                    } else {
                        request.deny()
                    }
                }
            }
        }

        // ========== STEP 5: JavaScript Interface for Voice Input ==========
        // Exposes native Android speech recognition to the web app.
        // Web Speech API (webkitSpeechRecognition) is NOT available inside WebView.
        // Instead, the web app calls window.AndroidBridge.startSpeechRecognition(),
        // which launches the Android system speech recognizer and returns results
        // via window.__onSpeechResult(transcript, error) callback.
        webView.addJavascriptInterface(object {
            @JavascriptInterface
            fun startSpeechRecognition() {
                runOnUiThread {
                    // Check RECORD_AUDIO permission (required on Android 6.0+)
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
                        ContextCompat.checkSelfPermission(
                            this@MainActivity,
                            Manifest.permission.RECORD_AUDIO
                        ) != PackageManager.PERMISSION_GRANTED
                    ) {
                        Log.d("[Voice]", "Requesting RECORD_AUDIO runtime permission")
                        pendingSpeechAfterPermission = true
                        permissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
                        return@runOnUiThread
                    }

                    launchSpeechRecognizer()
                }
            }
        }, "AndroidBridge")

        // ========== STEP 6: Load Your Web App ==========
        Log.d("[Dialer]", "Loading app: $APP_URL")
        webView.loadUrl(APP_URL)
    }

    /**
     * Launch the Android system speech recognizer intent.
     * Uses the modern ActivityResultLauncher API.
     */
    private fun launchSpeechRecognizer() {
        try {
            val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
                putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
                putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en-US")
                putExtra(RecognizerIntent.EXTRA_PROMPT, "Speak now...")
                putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 1)
            }
            speechLauncher.launch(intent)
            Log.d("[Voice]", "Speech recognition intent launched")
        } catch (e: Exception) {
            Log.e("[Voice]", "Speech recognition failed: ${e.message}")
            webView.evaluateJavascript("window.__onSpeechResult(null, 'not_available');", null)
            pendingSpeechAfterPermission = false
        }
    }

    /**
     * Enable immersive sticky mode — hides status bar and navigation bar.
     */
    private fun enableImmersiveMode() {
        @Suppress("DEPRECATION")
        window.addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN)
        window.decorView.systemUiVisibility = (
            View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY
                or View.SYSTEM_UI_FLAG_FULLSCREEN
                or View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
                or View.SYSTEM_UI_FLAG_LAYOUT_STABLE
                or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
                or View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
        )

        // Modern WindowInsetsController API (Android 11+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.setDecorFitsSystemWindows(false)
            WindowInsetsControllerCompat(window, window.decorView).let { controller ->
                controller.hide(WindowInsets.Type.statusBars() or WindowInsets.Type.navigationBars())
                controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            }
        }
    }

    /**
     * Re-apply immersive mode when activity resumes.
     */
    override fun onResume() {
        super.onResume()
        enableImmersiveMode()

        Log.d("[Dialer]", "onResume called. isCallInProgress=$isCallInProgress, phone=$currentCallPhone")

        if (isCallInProgress && currentCallPhone.isNotEmpty()) {
            val phone = currentCallPhone

            // Reset flags immediately
            isCallInProgress = false
            currentCallPhone = ""

            Log.d("[Dialer]", "User returned from dialer for: $phone — showing disposition popup")

            handler.postDelayed({
                triggerDispositionPopup(phone)
            }, 500)
        }
    }

    /**
     * Call JavaScript function in WebView to open the Post Call Disposition popup.
     */
    private fun triggerDispositionPopup(phoneNumber: String) {
        try {
            // Escape special characters to prevent JS injection
            val escaped = phoneNumber
                .replace("\\", "\\\\")
                .replace("'", "\\'")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
            val jsCode = """
                if (typeof window.showPostCallDisposition === 'function') {
                    window.showPostCallDisposition('$escaped');
                    true;
                } else {
                    console.log('showPostCallDisposition not found — saving to sessionStorage');
                    var data = { phone: '$escaped', timestamp: Date.now() };
                    sessionStorage.setItem('__trigger_disposition', JSON.stringify(data));
                    window.dispatchEvent(new CustomEvent('show-disposition-from-dialer', { detail: data }));
                    true;
                }
            """.trimIndent()

            webView.evaluateJavascript(jsCode) { result ->
                Log.d("[Dialer]", "evaluateJavascript result: $result")
            }
        } catch (e: Exception) {
            Log.e("[Dialer]", "evaluateJavascript FAILED: ${e.message}")
        }
    }

    // ========== HANDLE BACK BUTTON ==========
    @Deprecated("Deprecated in Java")
    override fun onBackPressed() {
        try {
            webView.evaluateJavascript("""
                (function() {
                    window.dispatchEvent(new CustomEvent('close-all-modals'));
                    var overlays = document.querySelectorAll('[data-slot="dialog-overlay"], [data-slot="alert-dialog-overlay"], .fixed.inset-0.bg-black');
                    if (overlays.length > 0) return true;
                    return false;
                })();
            """.trimIndent()) { result ->
                if (result == "true") {
                    Log.d("[Dialer]", "Closed modal via back button")
                } else {
                    if (webView.canGoBack()) {
                        webView.goBack()
                    } else {
                        @Suppress("DEPRECATION")
                        super.onBackPressed()
                    }
                }
            }
        } catch (e: Exception) {
            @Suppress("DEPRECATION")
            super.onBackPressed()
        }
    }

    // ========== CLEANUP ==========
    override fun onDestroy() {
        handler.removeCallbacksAndMessages(null)
        webView.destroy()
        super.onDestroy()
        Log.d("[Dialer]", "MainActivity destroyed")
    }
}
